code/RealMeService.php

```diff
@@ -205,58 +233,189 @@ class RealMeService extends Object
+        // First, check to see if we have an existing authenticated session
+        if($this->isAuthenticated()) {
+            return true;
+        }
+
+        // If not, attempt to retrieve authentication data from OneLogin (in case this is called during SAML assertion)
+        try {
+            $this->getAuth()->processResponse();
+            $errors = $this->getAuth()->getErrors();
+            $translatedMessage = null;
+
+            if(is_array($errors) && sizeof($errors) > 0) {
+                // The error message returned by onelogin/php-saml is the top-level error, but we want the actual error
+                if(isset($_POST) && isset($_POST['SAMLResponse'])) {
+                    $response = new OneLogin_Saml2_Response($this->getAuth()->getSettings(), $_POST['SAMLResponse']);
+                    $internalError = OneLogin_Saml2_Utils::query($response->document, "/samlp:Response/samlp:Status/samlp:StatusCode/samlp:StatusCode/@Value");
+
+                    if($internalError instanceof DOMNodeList && $internalError->length > 0) {
+                        $internalErrorCode = $internalError->item(0)->textContent;
+                        $translatedMessage = $this->findErrorMessageForCode($internalErrorCode);
+                    }
+                }
 
-        $auth->requireAuth(array(
-            'ReturnTo' => '/Security/realme/acs',
-            'ErrorURL' => '/Security/realme/error'
-        ));
+                // If we found a message to display, then let's redirect to the form and display it
+                if($translatedMessage) {
+                    $this->lastError = $translatedMessage;
+                    return false;
+                }
 
-        $loggedIn = false;
-        $authData = $this->getAuthData($auth);
+                SS_Log::log(
+                    sprintf(
+                        'onelogin/php-saml error messages: %s (%s)',
+                        join(', ', $errors),
+                        $this->getAuth()->getLastErrorReason()
+                    ),
+                    SS_Log::ERR
+                );
 
-        if (is_null($authData)) {
-            // no-op, $loggedIn stays false and no data is written
-        } else {
-            $this->config()->user_data = $authData;
-            Session::set('RealMeSessionDataSerialized', serialize($authData));
-            $loggedIn = true;
+                return false;
+            }
+
+            $authData = $this->getAuthData();
+
+            // If no data is found, then force login
+            if(is_null($authData)) {
+                throw new RealMeException('No SAML data, enforcing login', RealMeException::NOT_AUTHENTICATED);
+            }
+        } catch(Exception $e) {
+            // No auth data or failed to decrypt, enforce login again
+            $this->getAuth()->login(Director::absoluteBaseURL());
+            die;
+        }
+
+        $this->config()->user_data = $authData;
```

Comment:
Storing this data in the `Config` system doesn't seem like the right place for it. What are you trying to achieve by doing so?